go-controller/cmd/ovnkube/ovnkube.go

```diff
@@ -426,159 +426,174 @@ func runOvnKube(ctx context.Context, runMode *ovnkubeRunMode, ovnClientset *util
 		return ovnnode.CleanupClusterNode(runMode.identity)
 	}
 
-	stopChan := make(chan struct{})
-	wg := &sync.WaitGroup{}
-	defer func() {
-		close(stopChan)
-		wg.Wait()
-	}()
-
-	var masterWatchFactory *factory.WatchFactory
-
-	if runMode.ovnkubeController {
-		// create factory and start the controllers asked for
-		masterWatchFactory, err = factory.NewOVNKubeControllerWatchFactory(ovnClientset.GetOVNKubeControllerClientset())
-		if err != nil {
-			return err
-		}
-		defer masterWatchFactory.Shutdown()
+	watchFactory, err := newWatchFactory(runMode, ovnClientset)
+	if err != nil {
+		return fmt.Errorf("failed to initialize watch factory: %w", err)
 	}
 
+	// there might be dependencies accross components when starting so run them
+	// in separate threads
+	wg := &sync.WaitGroup{}
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithCancel(ctx)
+	var managerErr, controllerErr, nodeErr error
+
 	if runMode.clusterManager {
-		var clusterManagerWatchFactory *factory.WatchFactory
-		if runMode.ovnkubeController {
-			// if CM and NCM modes are enabled, then we should call the combo mode - NewMasterWatchFactory
-			masterWatchFactory, err = factory.NewMasterWatchFactory(ovnClientset.GetMasterClientset())
+		wg.Add(1)
+		go func() {
+			defer cancel()
```

Comment:
self-check:

1. try to start up CM
2. if all is well, we keep running
3. if start up fails then we decrement counter in WG, cancel the context which will trigger the channel close thus calling stop
4. after this we move to the next routine